content/blog/2025-06-06-a-plan-for-simd.md

```diff
@@ -0,0 +1,142 @@
+
+## Summary
+
+We build in the direction of [fearless_simd#2], but with a few course corrections.
+In particular, instead of manually curating the library and using (declarative) macros to try to reduce repetition and boilerplate, we rely heavily on code generation.
+
+While the current fearless\_simd\#2 prototype implements a handful of types, the goal is to support a full cross-product of SIMD width 64 to 512, signed and unsigned integer types from 8 to 32 (possibly 64), f32 (possibly f64), and f16 where available (primarily newer ARM chips, though Sapphire Rapids does support it).
+For widths greater than native, the library will polyfill with arrays of the native SIMD width.
+
+Note: since that blog post, f16 and Neon f16 instructions are supported in Rust nightly.
+It would be possible to polyfill in a hacky way (as is done in fearless\_simd\#2), but my inclination is to have these depend on Rust support, and similarly for AVX-512.
+With luck, these will stabilize soon (and obviously that’s one of our big asks of the Rust project).
+\[Update: AVX-512 will land in [1.89](https://github.com/rust-lang/rust/pull/138940) if things go well; still gathering more info about fp16 but that might also be very soon\]
+
+## Explicit vs variable width
+
+One of the big decisions in writing SIMD code is whether to write code with types of explicit width, or to use associated types in a trait which have chip-dependent widths.
+The design of pulp strongly favors the latter; the [Simd](https://docs.rs/pulp/latest/pulp/trait.Simd.html) trait has only “natural width” types, and, in particular, there is no implementation of, for example, 256 bit wide operations on Neon.
+A key departure from pulp, then, is support for portable explicit width programming.
+
+For Linebender work, I expect 256 bits to be a sweet spot.
+Obviously, it’s the natural width of AVX-2, which is a pretty big majority of x86\_64 chips.
+On Neon, code written for 256 will be unrolled somewhat, but there are 32 registers (as opposed to 16 for AVX-2), so register pressure should not be a problem, and high end chips (such as Apple Silicon) have very wide issue.
+The main reason to go smaller is when the loop has a high probability of consuming less than 256 bits of input.
+
+In the other direction, the majority of shipping AVX-512 chips are double-pumped, so code written to use 512 bits is not significantly faster (I assert this based on some serious experimentation on a Zen 5 laptop).
+I also expect this state of affairs to continue for a while; AMD won this design war and Intel is struggling to catch up after flailing for many years.
+If and when a significant number of true 512 bit chips ship, that would justify more work to write 512-optimized variants.
+
+For map-like workloads, I propose adding pulp-like associated natural-width types and methods to the Simd trait.
+Following Highway, I think it also makes sense to have some more methods that work in 128 bit blocks.
```

Comment:
This is the first (and only) mention of Highway here. Needs a link or something.